From 83462f050d5dd2ee18ffb09f1c8b80484f241fec Mon Sep 17 00:00:00 2001
From: Chelsio Cudbg <chelsiocudbg@users.noreply.github.com>
Date: Wed, 31 Dec 2025 13:21:43 -0500
Subject: [PATCH 4/5] misc fixes:

- fix port stats display
- fix resource leak in unload path
- fix port speed display
- fix pci remove related issues
---
 drivers/net/ethernet/chelsio/cxgb4/cxgb4_ethtool.c |  2 +-
 drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c    | 13 ++++++++++---
 drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c     |  6 ++----
 drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.h     |  2 ++
 4 files changed, 15 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_ethtool.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_ethtool.c
index 010a00688c04..6a2705b69a58 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_ethtool.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_ethtool.c
@@ -360,7 +360,7 @@ static void get_stats(struct net_device *dev, struct ethtool_stats *stats,
 	int i;
 	u64 *p0;
 
-	t4_get_port_stats_offset(adapter, pi->tx_chan,
+	t4_get_port_stats_offset(adapter, pi->lport,
 				 (struct port_stats *)data,
 				 &pi->stats_base);
 
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
index 20b49ab0b4ab..e0aca2b46ba0 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
@@ -311,6 +311,12 @@ static void link_report(struct net_device *dev)
 		case 100000:
 			s = "100Gbps";
 			break;
+		case 200000:
+			s = "200Gbps";
+			break;
+		case 400000:
+			s = "400Gbps";
+			break;
 		default:
 			pr_info("%s: unsupported speed: %d\n",
 				dev->name, p->link_cfg.speed);
@@ -2186,8 +2192,8 @@ static void attach_ulds(struct adapter *adap)
         list_add_tail(&adap->list_node, &adapter_list);
         for (i = 0; i < CXGB4_ULD_TYPE_MAX; i++) {
                 mutex_lock(&adap->uld_inst.uld_mutex);
-		mutex_unlock(&uld_mutex);
 		if (cxgb4_ulds[i].add) {
+			cxgb4_uld_alloc_resources(adap, i, &cxgb4_ulds[i]);
                         if (adap->flags & CXGB4_FULL_INIT_DONE)
                                 cxgb4_uld_txq_alloc_shared(adap, i);
                         uld_attach(adap, i);
@@ -2215,7 +2221,6 @@ static void detach_ulds(struct adapter *adap)
 					CXGB4_STATE_DETACH);
 			if (adap->flags & CXGB4_FULL_INIT_DONE)
 				cxgb4_uld_txq_free_shared(adap, i);
-			adap->uld_handle[i] = NULL;
 		}
 		mutex_unlock(&adap->uld_inst.uld_mutex);
 	}
@@ -6481,15 +6486,16 @@ void cxgb4_adap_remove(struct adapter *adapter)
 		 * valid filters ...
 		 */
 		cxgb4_filter_clear_all(adapter);
-		disable_interrupts(adapter);
 
 		if (cxgb4_uld_supported_any(adapter)) {
 			if (!list_empty(&adapter->list_node))
 				detach_ulds(adapter);
 			cxgb4_uld_queues_cleanup(adapter);
+			t4_uld_clean_up(adapter);
 		}
 		cxgb4_uld_cleanup(adapter);
 		adap_free_hma_mem(adapter);
+		disable_interrupts(adapter);
 
 		for_each_port(adapter, i)
 			if (adapter->port[i]->reg_state == NETREG_REGISTERED)
@@ -6562,6 +6568,7 @@ void cxgb4_adap_shutdown(struct adapter *adapter)
 		if (cxgb4_uld_supported_any(adapter)) {
 			detach_ulds(adapter);
 			cxgb4_uld_queues_cleanup(adapter);
+			t4_uld_clean_up(adapter);
 		}
 
 		disable_interrupts(adapter);
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
index 5402c7c4b2f5..bca2620aaaf2 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.c
@@ -2730,8 +2730,6 @@ void t4_uld_mem_free(struct adapter *adap)
 static void cxgb4_shutdown_uld_adapter(struct adapter *adap, enum cxgb4_uld_type type)
 {
 	if (adap->uld_handle[type]) {
-		adap->uld_handle[type]= NULL;
-		cxgb4_ulds[type].add = NULL;
 
 		if (adap->flags & CXGB4_FULL_INIT_DONE)
 			quiesce_rx_uld(adap, type);
@@ -2938,7 +2936,7 @@ int cxgb4_set_ktls_feature(struct adapter *adap, bool enable)
 #endif
 
 
-static void cxgb4_uld_alloc_resources(struct adapter *adap,
+void cxgb4_uld_alloc_resources(struct adapter *adap,
 				      enum cxgb4_uld_type type,
 				      const struct cxgb4_uld_info *p)
 {
@@ -3033,8 +3031,8 @@ int cxgb4_unregister_uld(enum cxgb4_uld_type type)
 		mutex_lock(&adap->uld_inst.uld_mutex);
 		if (adap->flags & CXGB4_FULL_INIT_DONE) {
 			cxgb4_uld_txq_free_shared(adap, type);
-			cxgb4_shutdown_uld_adapter(adap, type);
 		}
+		cxgb4_shutdown_uld_adapter(adap, type);
 		adap->uld_handle[type] = NULL;
 		mutex_unlock(&adap->uld_inst.uld_mutex);
 	}
diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.h b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.h
index 4ae82516bdbc..1eb352e2d39f 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.h
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_uld.h
@@ -586,6 +586,8 @@ unsigned int cxgb4_modparam_enable_ulds(void);
 int uld_attach(struct adapter *adap, unsigned int uld);
 void cxgb4_register_uld(enum cxgb4_uld_type type, const struct cxgb4_uld_info *p);
 int cxgb4_unregister_uld(enum cxgb4_uld_type type);
+void cxgb4_uld_alloc_resources(struct adapter *adap,
+			       enum cxgb4_uld_type type, const struct cxgb4_uld_info *p);
 bool cxgb4_uld_is_registered(struct adapter *adap, enum cxgb4_uld_type type);
 bool cxgb4_modparam_enable_ulds_supported(enum cxgb4_uld_type type);
 
-- 
2.39.1

